Add PAYGAS opcode

NIC619 opened the pull request:
### What was wrong?
Implementation of #234 


### How was it fixed?
#### Add `PAYGAS` opcode
* Carry additional informations such as `transaction_gas_limit`, `paygas_gasprice ` and `paygas_total_fee` in `ShardingMessage` object
    * `transaction_gas_limit` is the same as `transaction.gas`
    * `paygas_gasprice ` is set by `PAYGAS` opcode and only under two conditions:
        1. in a top level call(i.e., `computation.msg.depth == 0`)
        2. the value has not been set before during this transaction
    * If the two conditions are not met, current action is to do nothing.
    * `paygas_total_fee` is equal to `transaction_gas_limit * paygas_gasprice`
* Once triggered, deduct `paygas_total_fee` from transaction initiator(`transaction.to`)'s balance
* Refund fee is calculated base on `paygas_gasprice` instead of `transaction.gas_price`
* Opcode value is currently set as ~~`0x5c`~~ `0xf5`, placed in ~~Stack, Memory, Storage and Flow Operations~~ Systems section. I preferred to put it in Environment Information section but it's full already.

- [x] 1. Remove coinbase from `execute_transaction` and ~transfer the fee directly to `PAYGAS_TEMPORARY_ACCOUNT` of which the account data is enforced to be permanently stored in database hence it's witness data is not required.~ store the tx fee amount due in `block.tx_fee_sum` and transfer all fees to coinbase in `finalize_block` function
- ~2. Add a function that generates a transaction to transfer fees collected from all transactions of this block to coinbase.~
- 3. Abstract `gas_price` away, removing it from transaction format.(Moved to another PR)
- 4. Add tests.(More to be added in another PR)

#### Cute Animal Picture

![](http://www.strangetravel.com/images/content/136379.jpg)

NIC619 commented:
@vbuterin does the `PAYGAS_TEMPORARY_ACCOUNT` scheme in WIP item.1 and 2. makes sense?

pipermerriam commented on evm/logic/context.py:
```diff
@@ -193,3 +194,26 @@ def returndatacopy(computation):
+
+
+def paygas(computation):
+    gasprice = computation.stack.pop(type_hint=constants.UINT256)
```
*naming nitpick*

`gas_price` for conformity.

pipermerriam commented on evm/logic/context.py:
```diff
@@ -193,3 +194,26 @@ def returndatacopy(computation):
+
+    # Only valid if (1) triggered in a top level call and
+    # (2) not been set already during this transaction execution
+    if computation.msg.depth == 0 and computation.msg.paygas_gasprice == 0:
```
`computation.msg.paygas_gasprice == 0` seems problematic because it disallows a gas price of zero which at least historically, has been allowed.  Maybe use `None` as the *not-set* value?

pipermerriam commented on evm/vm/message.py:
```diff
@@ -187,12 +193,29 @@ def __init__(self,
+            return 0
+
+    @paygas_gasprice.setter
+    def paygas_gasprice(self, gas_price):
```
probably good to add validation for this value here to be **really** sure that it's valid.

pipermerriam commented on evm/vm/message.py:
```diff
@@ -187,12 +193,29 @@ def __init__(self,
+
+    @paygas_gasprice.setter
+    def paygas_gasprice(self, gas_price):
+        if self.paygas_gasprice == 0:
```
It seems like there should be a logic branch here which throws an exception (not a `VMException` but something that will bubble) if the value is already set, because if I understand correctly, we shouldn't even try to set the value if it's already set.

NIC619 replied:
What about returning `True` of `False` in `paygas_gasprice.setter` to indicate if the value is successfully set or not.
```python
if self.paygas_gasprice is None:
    self._paygas_gasprice = gas_price
    return True
else:
    return False
```

and in `PAYGAS` opcode check against this returned value and push `1` or `0` to the stack to indicate the outcome of this `PAYGAS` operation?
```python
def paygas(computation):
    ...
    if computation.msg.depth == 0 and computation.msg.paygas_gasprice(gas_price):
        …
        computation.stack.push(1)
    else:
        computation.stack.push(0)
```

It's like fail of value transfer in a call. It seems better to me that it does not result in failure of transaction but I'm not that confident about this. Could it be the situation that gas were paid in top level call but later during some calls to other contract, the contract trigger `PAYGAS` again?

vbuterin replied:
I like pushing 1 on success and 0 on failure.

pipermerriam replied:
So my suggestion was not focused on VM semantics, but rather our python API semantics.  Here's the situation I'm trying to avoid.

```python
# somewhere we set the gas price.
msg.paygas_gasprice = 10

# elsewhere in the code, we try to set it again.
msg.paygas_price = 100  # this will silently fail
```

What I would like is that the later case cannot silently fail, hence, suggesting that we raise an exception.  Then, if we ever have a location in the code which tries to set the gas price when it's already set which doesn't explicitely handle whatever exception we raise, then we'll get a hard failure in our code, rather than a silent failure that only manifests later at some point when the gas accounting is off.

pipermerriam replied:
@NIC619 

Look here: https://github.com/ethereum/py-evm/blob/master/evm/computation.py#L316-L338

Anything inside of vm execution which throws any exception that is a subclass of `VMError` gets handled there, resulting in `computation.is_error` being `True`.

What I'm proposing is something that throws something like a `TypeError` or another custom exception that **is not** a `VMError` subclass.  This way, if we write code which sets `msg.paygas_gasprice` without implementing error handling **and** the `msg.paygas_gasprice` has already been set... an error will be thrown in such a way that it won't be caught in the linked code above.

Does that make sense?

pipermerriam commented on evm/vm/message.py:
```diff
@@ -144,6 +144,8 @@ def is_create(self):
 class ShardingMessage(Message):
 
     is_create = False
+    transaction_gas_limit = None
```
Thoughts on removing the `transaction_` prefix here?  I'm not sure if my thinking is accurate, but historically, the `Message` object has been intentionally naive about the transaction.  Would it still be accurate if this was simply titled `gas_limit`?

Also, how does do `gas_limit` and `gas` differ?  Now that I'm asking makes me think that this is in-fact appropriately prefixed with `transaction_` assuming that's the difference, one is the message gas, the other is the outer transaction gas limit.

NIC619 replied:
Yes, you are correct. One is the outer transaction gas limit and one is message gas. And hence the prefix to tell the difference between them.

NIC619 commented:
@pipermerriam Made some updates. 

Now the exception `PAYGASResetGasprice` will be raised if trying to reset the already set `msg.paygas_gasprice`. And in `PAYGAS` opcode it would use try/except to check if `msg.paygas_gasprice` has been set already. Push `1` to the stack if not and `0` otherwise.

pipermerriam commented on evm/exceptions.py:
```diff
@@ -150,3 +150,7 @@ class UnannouncedStateAccess(VMError):
     pass
+
+
+class PAYGASResetGasprice(Exception):
```
other name suggestions:

- `GasPriceAlreadySet`
- `GasPriceOverwriteProtection`

I'm thinking the inclusing of `PAYGAS` in the name isn't needed and is a little visually confusing.

pipermerriam commented on evm/logic/context.py:
```diff
@@ -193,3 +195,30 @@ def returndatacopy(computation):
+
+    # Only valid if (1) triggered in a top level call and
+    # (2) not been set already during this transaction execution
+    if computation.msg.depth == 0:
```
The way this is written without an `else` clause this does nothing if triggered from `depth > 0`.  Is this intended?  Should it push `0` onto the stack if `depth > 0`?

NIC619 replied:
You are right. It should push `0` onto the stack if `depth > 0`.

pipermerriam commented on evm/logic/context.py:
```diff
@@ -193,3 +195,30 @@ def returndatacopy(computation):
+    # (2) not been set already during this transaction execution
+    if computation.msg.depth == 0:
+        try:
+            computation.msg.paygas_gasprice(gas_price)
```
This looks like an error.  You are calling a property which I believe should raise an exception stating that either `NoneType` or the integers are not callable.  

I do however like using a setter function rather than assigning to a property that throws an exception.  Maybe rename to something like `computation.msg.set_PAYGAS_gasprice(...)`?

NIC619 replied:
My mistake here. I will update the setter to `set_PAYGAS_gasprice` function.

pipermerriam replied:
Do you think it's worth implementing a few minimal test cases for this opcode to hit these code paths?  I know the JSON test suite will cover these eventually, but

- we'd probably benefit from having basic test cases for opcodes that were not buried in the hard to decipher JSON fixture tests.
- until we have JSON fixture tests for this it'd be nice to know it works (from a testing/CI perspective)

pipermerriam commented on evm/vm/message.py:
```diff
@@ -187,12 +196,34 @@ def __init__(self,
+            return self._paygas_gasprice
+
+    @paygas_gasprice.setter
+    def paygas_gasprice(self, gas_price):
```
Continuation of previous suggestion to remove the `setter` and just make this a method `set_PAYGAS_gasprice`.  Seems a bit more intuitive/correct/something.

pipermerriam commented on evm/vm/message.py:
```diff
@@ -187,12 +196,34 @@ def __init__(self,
+    @property
+    def paygas_gasprice(self):
+        if self._paygas_gasprice is None:
+            return 0
```
I'm still curious about returning `0` instead of `None` here.  It seems valuable to be able to distinguish between *not-set* and *set-to-0* which cannot be done in the current implementation.

NIC619 replied:
My rationale was that from inside(`Message` object) the two cases are distinguished by `msg._paygas_gasprice` and from outside one indeed can not tell the difference. But since when `msg.paygas_gasprice` is used to compute gas charge/refund (from outside) the two cases are basically the same.

Also if `msg.paygas_gasprice` could return `None`, we need to add a condition check every time it's used to compute gas charge/refund.
```python
gas_refund_amount = (gas_refund + gas_remaining) * computation.msg.paygas_gasprice if computation.msg.paygas_gasprice is not None else 0
```

pipermerriam replied:
So i'm on the fence about this.  Your call.  Here's some extra thoughts on why it *might* still be good to use `None`.

The `Computation` object has an existing convention of having `get_XXX` methods for things to account for these little bits of internal logic.  I think the following API is consisten with the rest of the class and clean.

```python
class Computation:
    ...
    def get_PAYGAS_gas_price(self):
        if self._paygas_gasprice is None:
            return 0
        else:
            return self._paygas_gasprice

# and then used elsewhere
gas_refund_amount = (gas_refund + gas_remaining) * computation.get_paygas_gasprice()
```

It also has the nice benefit of not reaching down into the message object.

Again, your call.  I'm +0 on this change if you don't think there is a valid use case for differentiating between set to 0 and unset.

NIC619 commented:
Since this had to be rebased and some part of this PR is already been taken cared of in other new PR, I just re-do it along with PR feedbacks.

I changed it back so that `paygas_gasprice` will return `None` if not set yet because the other case doesn't seem to have much advantage over this one.

I moved `paygas_garprice` and it's getter/setter to `Computation`. But at this moment of writing I feel like I might have misunderstood what you mean. @pipermerriam 
Do you mean (1)keep the `paygas_gasprice` in `Message` and use getter/setter in `Computation` to get/set this value, or (2)move `paygas_gasprice` and it's getter/setter together to `Computation`?

NIC619 commented:
Another thing is transferring the transaction fee to `coinbase`. I was thinking about hooking this information on to `computation` object(`computation.tx_fee`) since it's what's returned after executing transaction. Then add the fee information to `Block.tx_fee_sum` and finally transfer all the fees collected to `coinbase` in `finalize_block`.

```python
def execute_transaction(self, transaction):
    …
    transaction_fee = (transaction.gas - gas_remaining - gas_refund) * PAYGAS_gasprice
    computation.tx_fee = transaction_fee
    …

def add_transaction(self, transaction, computation, block):
    …
    if block.tx_fee_sum:
        block.tx_fee_sum += computation.tx_fee
    else:
        block.tx_fee_sum = computation.tx_fee
    …

def finalize_block(self, block):
    …
    with self.state_db() as state_db:
            state_db.delta_balance(block.header.coinbase, block.tx_fee_sum)
            self.logger.debug(
                "TOTAL TRANSACTON FEE: %s -> %s",
                block.tx_fee_sum,
                block.header.coinbase,
            )
    …
```

pipermerriam commented on evm/vm/forks/sharding/computation.py:
```diff
@@ -28,9 +29,24 @@
+    def get_PAYGAS_gas_price(self):
+        return self._paygas_gasprice
+
+    def set_PAYGAS_gasprice(self, gas_price):
```
I think that this is a good move (moving the value and the setters co `Computation`).  This function could use one extra validation check to enforce `depth == 0` within the setter itself.

NIC619 replied:
@pipermerriam 
Added the `depth == 0` validation check along with PAYGAS testing contract fixture and test suite.

hwwhww commented:
According to our last discussion, the `GAS_PAYGAS` is `3`, please update it.

pipermerriam commented on evm/vm/forks/sharding/vm_state.py:
```diff
@@ -160,14 +167,12 @@ def execute_transaction(self, transaction):
         )
-        with state_db_cm() as state_db:
-            state_db.delta_balance(self.coinbase, transaction_fee)
+        computation.tx_fee = transaction_fee
```
Rather than storing this value on the computation object what do you think about just re-calculating it within `add_transaction`?  It looks like we have all of the necessary information already available in the context of that function to do the computations.

NIC619 replied:
You are right. Totally makes sense.

pipermerriam commented on tests/core/helpers.py:
```diff
@@ -32,7 +32,8 @@ def new_sharding_transaction(
     """
-    tx_data = pad32(data_destination) + pad32(bytes([data_value])) + pad32(data_msgdata) + pad32(data_vrs)  # noqa: E501
+    assert len(data_vrs) in (0, 65)
+    tx_data = pad32(data_destination) + pad32(bytes([data_value])) + pad32(data_msgdata) + pad32(data_vrs[:32]) + pad32(data_vrs[32:64]) + bytes(data_vrs[64:])  # noqa: E501
```
suggest reformatting to 

```python
tx_data = (
    pad32(data_destination) +
    pad32(bytes([data_value])) +
    pad32(data_msgdata) +
    pad32(data_vrs[:32]) +
    pad32(data_vrs[32:64]) +
    bytes(data_vrs[64:]
)
```

NIC619 commented on evm/logic/system.py:
```diff
@@ -228,6 +230,7 @@ def __call__(self, computation):
             value=value,
             data=b'',
             code=call_data,
+            transaction_gas_limit=computation.msg.transaction_gas_limit,
```
Will do. 👍

pipermerriam commented on evm/logic/call.py:
```diff
@@ -342,3 +342,101 @@ def compute_msg_extra_gas(self, computation, gas, to, value):
+                    child_computation.output[:actual_output_size],
+                )
+
+            if not child_computation.should_burn_gas:
```
Note to self.

this is a confusing `if` statement because of the need for the `not`.  We might make a second property which returns the inverse of this as `computation.should_return_gas`.

NIC619 replied:
Noted. Will make a PR to update this.

pipermerriam commented on evm/vm/forks/sharding/vm_state.py:
```diff
@@ -185,3 +189,97 @@ def execute_transaction(self, transaction):
+        gas_remaining = computation.get_gas_remaining()
+        gas_refunded = computation.get_gas_refund()
+        gas_used = transaction.gas - gas_remaining
+        gas_refund = min(gas_refunded, gas_used // 2)
```
Hrm, rather than repeating this block of code both here and above, I think we should dig it out into a utility of some sort.

```python
def compute_transaction_gas_amounts(computation, transaction_gas, gas_price):
    gas_remaining = computation.get_gas_remaining()
    gas_refunded = computation.get_gas_refund()
    gas_used = transaction_gas - gas_remaining
    gas_refund = min(gas_refunded, gas_used // 2)
    tx_fee = (transaction_gas - gas_remaining - gas_refund) * gas_price
    return tx_fee, gas_used, gas_refund
```

Something like this.  Could also be implemented as a method on the `Computation` class which would allow removing the `gas_price` argument from the call signature.

pipermerriam commented on evm/vm/forks/sharding/vm_state.py:
```diff
@@ -185,3 +189,97 @@ def execute_transaction(self, transaction):
+        tx_fee = (transaction.gas - gas_remaining - gas_refund) * tx_PAYGAS_gasprice
+        # Bookkeep this transaction fee
+        if hasattr(block, "tx_fee_sum"):
+            block.tx_fee_sum += tx_fee
```
I'd like to figure out a way to do this which doesn't involve storing data on the block object.  Looking around at the code however, it's not very clear where that would be.

For now, lets just formalize this property on the block object so that we don't have to do `hasattr` checks.
